crypto/crypto/src/noise.rs

```diff
@@ -0,0 +1,587 @@
+    #[cfg(test)]
+    return sha2::Sha256::digest(data).to_vec();
+}
+
+fn hkdf(ck: &[u8], dh_output: Option<&[u8]>) -> Result<(Vec<u8>, Vec<u8>), NoiseError> {
+    let dh_output = dh_output.unwrap_or_else(|| &[]);
+    #[cfg(not(test))]
+    let hkdf_output = Hkdf::<Sha3_256>::extract_then_expand(Some(ck), dh_output, None, 64);
+    #[cfg(test)]
+    let hkdf_output = Hkdf::<sha2::Sha256>::extract_then_expand(Some(ck), dh_output, None, 64);
+
+    let hkdf_output = hkdf_output.map_err(|_| NoiseError::Hkdf)?;
+    let (k1, k2) = hkdf_output.split_at(32);
+    Ok((k1.to_vec(), k2.to_vec()))
+}
+
+fn mix_hash(h: &mut Vec<u8>, data: &[u8]) {
+    h.extend_from_slice(data);
+    *h = hash(h);
+}
+
+fn mix_key(ck: &mut Vec<u8>, dh_output: &[u8]) -> Result<Vec<u8>, NoiseError> {
+    let (new_ck, mut k) = hkdf(ck, Some(dh_output))?;
+    *ck = new_ck;
+    if !cfg!(test) {
+        k.resize(16, 0);
+    }
+    Ok(k)
+}
+
+fn checked_vec_to_public_key(bytes: &[u8]) -> Result<x25519::PublicKey, NoiseError> {
+    if bytes.len() != X25519_PUBLIC_KEY_LENGTH {
+        return Err(NoiseError::WrongPublicKeyReceived);
+    }
+    let mut array = [0u8; X25519_PUBLIC_KEY_LENGTH];
+    array.copy_from_slice(bytes);
+    Ok(array.into())
+}
+
+// const
+
+/// We only implement the Noise_IK_25519_AES128GCM_SHA3 protocol, hence we can pre-compute the hash
+pub static PROTOCOL_NAME: once_cell::sync::Lazy<Vec<u8>> = once_cell::sync::Lazy::new(|| {
+    if cfg!(test) {
+        b"Noise_IK_25519_AESGCM_SHA256\0\0\0\0".to_vec()
+    } else {
+        // only hash the protocol name if it's greater than HASHLEN
+        hash(b"Noise_IK_25519_AES128GCM_SHA3-256")
+    }
+});
+
+//
+// IK:
+//  <- s
+//  ...
+//  -> e, es, s, ss
+//  <- e, ee, se
+//
+
+//
+// Initiator
+// ---------
+//
+
+/// A key holder structure used for both initiators and responders.
+pub struct NoiseConfig {
+    private_key: x25519::StaticSecret,
+    public_key: [u8; X25519_PUBLIC_KEY_LENGTH],
+}
+
+/// Refer to the Noise protocol framework specification in order to understand these fields.
+pub struct InitiatorHandshakeState {
+    /// rolling hash
+    h: Vec<u8>,
+    /// chaining key
+    ck: Vec<u8>,
+    /// ephemeral key
+    e: x25519::StaticSecret,
+}
+
+impl NoiseConfig {
+    /// A peer must create a NoiseConfig through this function before being able to connect with other peers.
+    pub fn new(private_key: x25519::StaticSecret) -> Self {
+        let public_key = x25519::PublicKey::from(&private_key);
+        Self {
+            private_key,
+            public_key: public_key.as_bytes().to_owned(),
+        }
+    }
+
+    /// An initiator can use this function to initiate a handshake with a known responder.
+    pub fn initiate_connection(
+        &self,
+        rng: &mut (impl rand::RngCore + rand::CryptoRng),
+        prologue: &[u8],
+        remote_public: &x25519::PublicKey,
+        payload: Option<&[u8]>,
+    ) -> Result<(InitiatorHandshakeState, Vec<u8>), NoiseError> {
```

Comment:
instead of allocating a new `Vec<u8>` for msg, the caller should be able to pass in a  `msg_buffer: &mut [u8]` and initiate_connection should write directly into that (probably just wrap it in a `Cursor` in the method). In fact, all handshake and write/read_message APIs should follow this pattern. Ideally, this noise crate should almost never allocate and allow the caller to control allocations. Unfortunately the rest of the crypto/ APIs are not very good about this. This will also mean we can easily reuse `netcore::NoiseSocket`

Reply:
> Ideally, this noise crate should almost never allocate and allow the caller to control allocations

I'll do this next! good point